frontend/app/src/components/StatusWidget/StatusWidget.tsx

```diff
@@ -136,237 +106,244 @@ const RUNNING_MAN_DISPLAY_DELAY_TIME_MS = 500
+
+  const handleSessionEvent = useCallback(
+    (event: SessionEvent): void => {
+      if (event.type === "scriptChangedOnDisk") {
+        setScriptChangedOnDisk(true)
+        setPromptMinimized(false)
+        minimizePromptAfterTimeout(PROMPT_DISPLAY_INITIAL_TIMEOUT_MS)
+      }
+    },
+    [minimizePromptAfterTimeout]
+  )
+
+  const showRunningManAfterInitialDelay = useCallback(
+    (delay: number): void => {
+      if (showRunningManTimer.current !== null) {
+        showRunningManTimer.current.setTimeout(() => {
+          setShowRunningMan(true)
+        }, delay)
+      }
+    },
+    []
+  )
 
-  handleKeyDown = (keyName: string): void => {
-    // NOTE: 'r' is handled at the App Level
-    if (keyName === "a") {
-      this.handleAlwaysRerunClick()
-    }
-  }
+  const handleScroll = useCallback((): void => {
+    setStatusMinimized(shouldMinimize())
+  }, [])
 
-  /** Called by React on prop changes */
-  public static getDerivedStateFromProps(
-    props: Readonly<StatusWidgetProps>
-  ): Partial<State> | null {
-    // Reset transient event-related state when prop changes
-    // render that state irrelevant
-    if (props.scriptRunState === ScriptRunState.RUNNING) {
-      return { scriptChangedOnDisk: false, promptHovered: false }
-    }
+  const onAppPromptHover = useCallback((): void => {
+    setPromptHovered(true)
+  }, [])
 
-    return null
-  }
+  const onAppPromptUnhover = useCallback((): void => {
+    setPromptHovered(false)
+    setPromptMinimized(false)
+    minimizePromptAfterTimeout(PROMPT_DISPLAY_HOVER_TIMEOUT_MS)
+  }, [minimizePromptAfterTimeout])
 
-  public componentDidMount(): void {
-    this.sessionEventConn =
-      this.props.sessionEventDispatcher.onSessionEvent.connect(e =>
-        this.handleSessionEvent(e)
-      )
-    window.addEventListener("scroll", this.handleScroll)
-  }
+  const handleStopScriptClick = useCallback((): void => {
+    stopScript()
+  }, [stopScript])
 
-  public componentWillUnmount(): void {
-    if (this.sessionEventConn !== undefined) {
-      this.sessionEventConn.disconnect()
-      this.sessionEventConn = undefined
-    }
+  const handleRerunClick = useCallback((): void => {
+    rerunScript(false)
+  }, [rerunScript])
 
-    this.minimizePromptTimer.cancel()
-    this.showRunningManTimer.cancel()
+  const isNewYears = useCallback((): boolean => {
+    // Test if current date between 12/31 & 1/06
+    const currentDate = new Date()
+    const month = currentDate.getMonth()
+    const date = currentDate.getDate()
+    // Check if Dec 31st
+    if (month === 11 && date === 31) return true
+    // Check if Jan 1st through 6th
+    if (month === 0 && date <= 6) return true
+    return false
+  }, [])
 
-    window.removeEventListener("scroll", this.handleScroll)
+  function shouldMinimize(): boolean {
+    return window.scrollY > 32
   }
 
-  private isConnected(): boolean {
-    return this.props.connectionState === ConnectionState.CONNECTED
+  function promptButton(
```

Comment:
nit: a function that returns jsx sounds like a good candidate for a component :D

Reply:
[484e7e7](https://github.com/streamlit/streamlit/pull/9851/commits/484e7e7955e0844996fc2979ef01128bde97821e)